### PR TITLE
feat: add token

### PR DESCRIPTION
**Pull request by Butcher-L**
apply token and add test 
add product route
add addition field in adding product and add test

### CODE REVIEW

File: source/controllers/products.ts [hunk 1]
```diff
@@ -1,6 +1,49 @@
-import Mongoose from 'mongoose';
 import Products from '../models/products';
 
+const addProducts = async(ctx,next) =>{
+    try {
```
Comment (ccpacillos) [inline source/controllers/products.ts:4]: No need to `try-catch` here.

File: source/controllers/products.ts [hunk 2]
```diff
@@ -1,6 +1,49 @@
...
+    let { name, price ,createdBy} = ctx.request.body;
```
Comment (ccpacillos) [inline source/controllers/products.ts:5]: There's no `createdBy` payload input in creating a product. Follow the specifications in the exercise.

File: source/controllers/products.ts [hunk 3]
```diff
@@ -1,6 +1,49 @@
-import Mongoose from 'mongoose';
 import Products from '../models/products';
 
+const addProducts = async(ctx,next) =>{
```
Comment (ccpacillos) [inline source/controllers/products.ts:3]: I think you can export these methods immedaitely.

`export const addProducts = async (ctx, next) => {}`

or, if you don't want to use arrow function:

`export async function addProducts() {}`

File: source/controllers/products.ts [hunk 4]
```diff
@@ -1,6 +1,49 @@
...
+        ctx.message='Price require'
+        return
+    }
+    if(!createdBy){
+        ctx.status= 400
+        ctx.message='Owner require'
+        return
+    }
+
+    const find = await Products.exists({name})
```
Comment (ccpacillos) [inline source/controllers/products.ts:29]: Uniqueness of the product should be based on the input `id`.

File: source/controllers/products.ts [hunk 5]
```diff
@@ -1,6 +1,49 @@
...
+    if(find){
+        ctx.status = 400
+        ctx.body='Product already exist'
+    }else{
+        const res = await product.save()
+    
+        ctx.status=200
+        ctx.body={res}
+    }
+    } catch (error) {
```
Comment (ccpacillos) [inline source/controllers/products.ts:40]: Observe proper indentations.

File: source/controllers/users.ts [hunk 6]
```diff
@@ -1,17 +1,22 @@
...
 import Users from '../models/users';
-import SignJwt from '../functions/sign-jwt';
 import Config from '../configs/config'
 import Jwt from 'jsonwebtoken'
 
-
-
-
 const addUser = async(ctx ,next) => {
     let { username, password } = ctx.request.body;
```
Comment (ccpacillos) [inline source/controllers/users.ts:7]: Use the basic auth header to get the `username` and `password` pair.

File: source/controllers/users.ts [hunk 7]
```diff
@@ -1,17 +1,22 @@
...
+        if(!username){
+            ctx.status= 400
+            ctx.message='Username require'
+            return
+        }
+        if(!password){
+            ctx.status= 400
+            ctx.message='Password require'
+            return
+        }
```
Comment (ccpacillos) [inline source/controllers/users.ts:18]: This code change seems to be another scope.
So far, you have the scopes:
- add product
- add user

Create a separate PR for each scope for much easier review process.

File: source/middleware/extract-jwt.ts [hunk 8]
```diff
@@ -1,26 +1,37 @@
...
+  try {
+    decoded = Jwt.verify(token, Config.token.secret);
+  } catch (e) {
+    ctx.status=400
+    ctx.body='Session expired'
+    allowed = false;
+    return
+  }
+await next()  
+}
```
Comment (ccpacillos) [inline source/middleware/extract-jwt.ts:34]: This is another scope again. Please put this in a separate PR.

**Review by ccpacillos — CHANGES_REQUESTED** [review]

File: source/controllers/products.ts (hunk 4 above)
Comment (Butcher-L) [reply source/controllers/products.ts:29]: this is to verify the product name is existing, is it okay to have multiple product names?

File: source/controllers/products.ts (hunk 3 above)
Comment (Butcher-L) [reply source/controllers/products.ts:3]: do I need to separate the function per route? like the addProducts in source/controllers/products and make it in another file?

Comment (Butcher-L) [reply source/controllers/products.ts:3]: can i create a new PR for this?
